src/gax/src/retry_policy.rs

```diff
@@ -0,0 +1,954 @@
+// Copyright 2024 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Defines traits for retry policies and some common implementations.
+//!
+//! The client libraries automatically retry RPCs when (1) they fail due to
+//! transient errors **and** the RPC is [idempotent], (2) or failed before an
+//! RPC was started. That is, when it is safe to attempt the RPC more than once.
+//!
+//! Applications may override the default behavior, increasing the retry
+//! attempts, or changing what errors are considered safe to retry.
+//!
+//! This module defines the traits for retry policies and some common
+//! implementations.
+//!
+//! # Example:
+//! ```
+//! # use std::sync::Arc;
+//! # use gcp_sdk_gax::retry_policy::*;
+//! # use gcp_sdk_gax::options::RequestOptionsBuilder;
+//! fn customize_retry_policy(builder: impl RequestOptionsBuilder) -> impl RequestOptionsBuilder {
+//!     builder.with_retry_policy(LimitedAttemptCount::provider(5))
+//! }
+//! ```
+//!
+//! [idempotent]: https://en.wikipedia.org/wiki/Idempotence
+
+use crate::error::rpc::Status;
+use crate::error::{Error, HttpError};
+use std::cell::Cell;
+use std::sync::Arc;
+use std::sync::Mutex;
+
+/// The result of a retry policy decision.
+///
+/// If the caller should continue retrying the policy returns
+/// [Continue][std::ops::ControlFlow::Continue]. If the caller
+/// should stop retrying, the policy returns
+/// [Break][std::ops::ControlFlow::Break].
+///
+/// In both variants the result includes an error. This is useful when retry
+/// policies are composed. The inner policy returns `Continue` based on the
+/// error type, and the outer policy may return `Break` based on the number
+/// errors, or the elapsed time.
+#[derive(Debug)]
+pub enum RetryFlow {
```

Comment:
nit: I wonder if this should be called RetryFlowFailure or something as all of these are the sad path of a retry loop.

Reply:
Maybe?  This was inspired by `std::ops::ControlFlow`, but I am terrible at naming.

With that said, is not all sad path: `RetryFlow::Continue` is the happy case.